Part 1 intro to R (#5)
carpentries-incubator/r-geospatial-urban: episodes/02-data-structures.Rmd

---
title: 'Data Structures'
teaching: 10
exercises: 2
---

:::::::::::::::::::::::::::::::::::::: questions 

- What are the basic data types in R?
- How do I represent categorical information in R?

::::::::::::::::::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::: objectives

- To be aware of the different types of data.
- To begin exploring data frames, and understand how they are related to vectors, factors and lists.
- To be able to ask questions from R about the type, class, and structure of an object.

::::::::::::::::::::::::::::::::::::::::::::::::

> ## Vectors 

Author: Done in a111918. See lines 44-60

> ## Vectors 

Reviewer: line 58: I dont think this is the right place. Vector data vs vector data structure is quite confusing when you have never heard of vector vs raster layers. I think this should be included when we explain the difference between vector and raster.

> ## Vectors 

Author: I agree, I changed the text accordingly. As soon as we introduce geospatial vector data, we can refer back to disambiguate. I can try to do that at the beginning of the vector lesson.

It can be:

- `numerical` (a number)
- `integer` (a number without information about decimal points)
- `logical` (a boolean - are values TRUE or FALSE?)
- `character` (a text/ string of characters)
- `complex` (a complex number)
- `raw` (raw bytes)

We won't discuss `complex` or `raw` data type in the workshop.

:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::

You can create a vector with a `c()` function. 

```{r vectors}

numeric_vector <- c(2, 6, 3) # vector of numbers - numeric data type.
numeric_vector

character_vector <- c('banana', 'apple', 'orange') # vector of words - or strings of characters- character data type
character_vector

logical_vector <- c(TRUE, FALSE, TRUE) # vector of logical values (is something true or false?)- logical data type.
logical_vector

```

### Combining vectors 

The combine function, `c()`, will also append things to an existing vector:

```{r combine-vectors}

ab_vector <- c('a', 'b')
ab_vector

abcd_vector <- c(ab_vector, 'c', 'd')
abcd_vector

```

### Missing values  

A common operation you want to perform is to remove all the missing values 
(in R denoted as `NA`). Let's have a look how to do it: 

> 

Reviewer 2: Let the participants create the vector as a mini-challenge in the sake of interactivity?

Reviewer: ok

Author: @KyriJanssen, @fcjerome, can you please update the text with a challenge. Again, you can do that as a suggestion using this button:
![image](https://github.com/carpentries-incubator/r-geospatial-urban/assets/33600128/a60b7e23-7cac-4a33-bfc9-e0238a589b9a)

You can select multiple lines before making suggestions. See how it's done here: https://stackoverflow.com/questions/48911087/how-to-select-comment-on-a-range-of-lines-in-github-pull-request

However, sometimes, you would like to have the `NA` 
permanently removed from your vector. 
For this you need to identify which elements of the vector hold missing values 
with `is.na()` function. 

```{r remove-na2}
is.na(with_na) #  This will produce a vector of logical values, stating if a statement 'This element of the vector is a missing value' is true or not

!is.na(with_na) # # The ! operator means negation ,i.e. not is.na(with_na)

```

We know which elements in the vectors are `NA`.
Now we need to retrieve the subset of the `with_na` vector that is not `NA`.
Sub-setting in `R` is done with square brackets`[ ]`. 

```{r remove-na3}

without_na <- with_na[ !is.na(with_na) ] # this notation will return only the elements that have TRUE on their respective positions

without_na

```

## Factors 

Another important data structure is called a **factor**. 
Factors look like character data, but are used to represent categorical information.

Factors create a structured relation between the different levels (values) of a
categorical variable, such as days of the week or responses to a question in a
survey. While factors look (and often behave) like character vectors, they are 

> actually treated as numbers by `R`. 

Author: ```suggestion
actually treated as numbers by `R`, which is useful for computing summary 
statistics about their distribution, running regression analysis, etc.
```

> actually treated as numbers by `R`. 

Author: @ClementineCttn feel free to add direct suggestions

### Create factors
Once created, factors can only contain a pre-defined set of values, 
known as levels. 

```{r factor-create}

nordic_str <- c('Norway', 'Sweden', 'Norway', 'Denmark', 'Sweden')
nordic_str # regular character vectors printed out

nordic_cat <- factor(nordic_str) # factor() function converts a vector to factor data type
nordic_cat # With factors, R prints out additional information - 'Levels'

```

### Inspect factors
R will treat each unique value from a factor vector as a **level** and (silently)
assign numerical values to it. 
This can come in handy when performing statistical analysis. 
You can inspect and adapt levels of the factor. 

```{r factor-inspect}
levels(nordic_cat) # returns all levels of a factor vector.  

nlevels(nordic_cat) # returns number of levels in a vector
```

### Reorder levels
Note that `R` sorts the levels in the alphabetic order, 
not in the order of occurrence in the vector. `R` assigns value of:

- 1 to level 'Denmark',
- 2 to 'Norway' 
- 3 to 'Sweden'.

This is important as it can affect e.g. the order in which categories are 
displayed in a plot or which category is taken as a baseline in a statistical model.

> You can reorder the categories using `factor()` function.

Reviewer 2: This can be useful to select a reference category (first level) in a regression model, rather than using the default category systematically (i.e. based on alphabetical order)

```{r factor-reorder}
nordic_cat <- factor(nordic_cat, levels = c('Norway' , 'Denmark', 'Sweden')) # now Norway should be the first category, Denmark second and Sweden third

nordic_cat
```

:::::::::::::::::::::::::::::::::::::::::::::::::::::: callout 
There is more than one way to reorder factors. Later in the lesson,
we will use `fct_relevel()` function from `forcats` package to do the reordering.

> ```{r factor-reorder}

Author: Indeed code chunk names can cause that issue. Renaming them sequentially or removing names entirely are also options.

### Create factors
Once created, factors can only contain a pre-defined set of values, 
known as levels. 

```{r factor-create}

nordic_str <- c('Norway', 'Sweden', 'Norway', 'Denmark', 'Sweden')
nordic_str # regular character vectors printed out

nordic_cat <- factor(nordic_str) # factor() function converts a vector to factor data type
nordic_cat # With factors, R prints out additional information - 'Levels'

```

### Inspect factors
R will treat each unique value from a factor vector as a **level** and (silently)
assign numerical values to it. 
This can come in handy when performing statistical analysis. 
You can inspect and adapt levels of the factor. 

```{r factor-inspect}
levels(nordic_cat) # returns all levels of a factor vector.  

nlevels(nordic_cat) # returns number of levels in a vector
```

### Reorder levels
Note that `R` sorts the levels in the alphabetic order, 
not in the order of occurrence in the vector. `R` assigns value of:

- 1 to level 'Denmark',
- 2 to 'Norway' 
- 3 to 'Sweden'.

This is important as it can affect e.g. the order in which categories are 
displayed in a plot or which category is taken as a baseline in a statistical model.

> You can reorder the categories using `factor()` function.

Author: ```suggestion
You can reorder the categories using `factor()` function. This can be useful, for instance, to select a reference category (first level) in a regression model or for ordering legend items in a plot, rather than using the default category systematically (i.e. based on alphabetical order).
```

Data type refers to a type of information that is stored by a value. 
It can be:

- `numerical` (a number)
- `integer` (a number without information about decimal points)
- `logical` (a boolean - are values TRUE or FALSE?)
- `character` (a text/ string of characters)
- `complex` (a complex number)
- `raw` (raw bytes)

We won't discuss `complex` or `raw` data type in the workshop.

:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::

:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: callout

### Data structures

Vectors are the most common and basic data structure in R but you will come 
across other data structures such as data frames, lists and matrices as well.
In short:

- data.frames is a two-dimensional data structure in which columns are vectors of the same length that can have different data types. We will use this data structure in this lesson.
- lists can have an arbitrary structure and can mix data types;
- matrices are two-dimensional data structures containing elements of the same data type.

For a more detailed description, see [Data Types and Structures](https://swcarpentry.github.io/r-novice-inflammation/13-supp-data-structures.html).

> Data structures for geospatial data use combinations of lists, data frames, matrices and vectors. While vector data (not the data structure with the same name!) combines a data frame structure with lists for attributes and geometry in a nested way, raster data are essentially represented as matrices in which each cell is a pixel. More about vector and raster data later in the workshop!

Author: ```suggestion
Note that vector data in the geospatial context is different from vector data types. More about vector data in a [later lesson](../episodes/09-open-and-plot-vector-layers.Rmd)!
```

:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::

You can create a vector with a `c()` function. 

```{r vectors}

numeric_vector <- c(2, 6, 3) # vector of numbers - numeric data type.
numeric_vector

character_vector <- c('banana', 'apple', 'orange') # vector of words - or strings of characters- character data type
character_vector

logical_vector <- c(TRUE, FALSE, TRUE) # vector of logical values (is something true or false?)- logical data type.
logical_vector

```

### Combining vectors 

The combine function, `c()`, will also append things to an existing vector:

```{r combine-vectors}

ab_vector <- c('a', 'b')
ab_vector

abcd_vector <- c(ab_vector, 'c', 'd')
abcd_vector

```

> 
> ### Missing values  
> 

Reviewer: ```suggestion

### Exercise 
<div class="alert alert-info">
<strong>Challenge</strong> 
Combine the abcd_vector with the numerical_vector in R. What is the datatype of this new vector and why?

### Missing values  

```

Author: Thanks, Kyri. I transformed this into a challenge with a hidden solution.

:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::: callout

### Data structures

Vectors are the most common and basic data structure in R but you will come 
across other data structures such as data frames, lists and matrices as well.
In short:

- data.frames is a two-dimensional data structure in which columns are vectors of the same length that can have different data types. We will use this data structure in this lesson.
- lists can have an arbitrary structure and can mix data types;
- matrices are two-dimensional data structures containing elements of the same data type.

For a more detailed description, see [Data Types and Structures](https://swcarpentry.github.io/r-novice-inflammation/13-supp-data-structures.html).

Note that vector data in the geospatial context is different from vector data types. More about vector data in a [later lesson](../episodes/09-open-and-plot-vector-layers.Rmd)!

:::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::

You can create a vector with a `c()` function. 

```{r vectors}

numeric_vector <- c(2, 6, 3) # vector of numbers - numeric data type.
numeric_vector

> character_vector <- c('banana', 'apple', 'orange') # vector of words - or strings of characters- character data type

Author: Use city names here
